### PR TITLE
apiserver: refactor PolicyRuleEvaluator interface

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

We are proposing a change to `Policy` API to make room for a boolean field that dictates whether the managed fields should be omitted from the audit event.  This is the change we are proposing:
```
type Policy struct {
	// +optional
	OmitManagedFields bool `json:"omitManagedFields,omitempty"`
}

type PolicyRule struct {
	// +optional
	OmitManagedFields *bool `json:"omitManagedFields,omitempty"`
}
```

it's in both `Policy` and `PolicyRule` so the cluster operator has an option to use it in a more granular way:
- the `bool` in `Policy` dictates the global default policy.
- the `*bool` in `PolicyRule` can override the global default for a particular matching rule, with three interpretations:
  - `nil` will represent 'not specified by the operator', hence use the global default.
  - `{false|true}` will override the global default.

To accommodate this, we need to refactor `PolicyRuleEvaluator` interface so it returns a `struct` into which we can compose the `OmitManagedFields` boolean value from the Policy object.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/audit/context.go [hunk 1]
```diff
@@ -82,3 +87,23 @@ func auditAnnotationsFrom(ctx context.Context) []annotation {
 
 	return *annotations
 }
+
+// WithAuditEvent returns set audit event struct.
+func WithAuditEvent(parent context.Context, ev *EvaluatedPolicyAndEvent) context.Context {
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/context.go:92]: the move was necessary to avoid cyclic import

**Comment by tkashem** [comment]
This will pave the way for https://github.com/kubernetes/kubernetes/pull/94986

File: staging/src/k8s.io/apiserver/pkg/audit/evaluator.go [hunk 2]
```diff
@@ -33,14 +33,28 @@ type EventAndEvaluatedPolicyPair struct {
...
 }
 
-// EvaluatedPolicy is the evaluated policy that is applicable to a given
-// a request. PolicyRuleEvaluator evaluates the audit configuration for a
-// request against the authorizer attributes and returns an EvaluatedPolicy
+// EvaluatedPolicy is the evaluated policy that is applicable to a given request.
+// Note: we intentionally exclude the Level at which the request is being
+// audited since it is already included in the audit Event struct. There are
+// many code sites that refer to 'Event.Level'.
+type EvaluatedPolicy struct {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/evaluator.go:40]: this is more a RequestAuditConfig

File: staging/src/k8s.io/apiserver/pkg/audit/evaluator.go [hunk 3]
```diff
@@ -21,6 +21,24 @@ import (
 	"k8s.io/apiserver/pkg/authorization/authorizer"
 )
 
+// EventAndEvaluatedPolicyPair is a pair of the evaluated policy that applies
+// to a given request and the audit Event object that is bein captured. It's
+// a convenient placeholder to store both these objects in the request context.
+type EventAndEvaluatedPolicyPair struct {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/evaluator.go:27]: maybe AuditContext ?

File: staging/src/k8s.io/apiserver/pkg/audit/evaluator.go [hunk 4]
```diff
@@ -36,10 +36,24 @@ type AuditContext struct {
 // RequestAuditConfig is the evaluated audit configuration that is applicable to
 // a given request. PolicyRuleEvaluator evaluates the audit policy against the
 // authorizer attributes and returns a RequestAuditConfig that applies to the request.
-type RequestAuditConfig struct{}
+type RequestAuditConfig struct {
+	// OmitStages is the stages that need to be omitted from being audited.
+	OmitStages []audit.Stage
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/audit/evaluator.go:41]: `OmitManagedFields` will be added here in https://github.com/kubernetes/kubernetes/pull/94986

File: staging/src/k8s.io/apiserver/pkg/audit/evaluator.go [hunk 5]
```diff
@@ -21,8 +21,39 @@ import (
...
+
+	// Level at which the request is being audited at
+	Level audit.Level
+}
+
 // PolicyRuleEvaluator exposes methods for evaluating the policy rules.
 type PolicyRuleEvaluator interface {
 	// Check the audit level for a request with the given authorizer attributes.
-	LevelAndStages(authorizer.Attributes) (audit.Level, []audit.Stage)
+	LevelAndStages(authorizer.Attributes) RequestAuditConfigWithLevel
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/evaluator.go:58]: the func name is misleading now. Maybe just `EvaluatePolicyRule`?

**Comment by sttts** [comment]
Looks good. One comment left.

**Comment by tkashem** [comment]
/assign  @dchen1107 @yguo0905 @cheftako

needs an approval from an approver of gce. it's a simple refactor of audit interface.

**Comment by tkashem** [comment]
gentle ping to @dchen1107 @yguo0905 @cheftako for an approval - it's a simple refactor.

**Pull request merged**